### PR TITLE
Add hp to FEInterfaceValues

### PR DESCRIPTION
**Pull request by fdrmrc**
This allows one to construct and use FEInterfaceValues with `hp::{Mapping,FE,Q}Collection`.
In practice, I mimicked the non-hp case by adding members like `internal_hp_fe_face_values`, `internal_hp_fe_subface_values` (and so on) to which I can query `get_present_fe_values(cell,face_no)`

In order to avoid another if clause in the `reinit`, I added a `reinit_hp`. I don't think this approach is satisfactory, due to the fact that a user is required to know that the "reinit" name is different in this context. Maybe a `hp::FEInterfaceValues` could be a good workaround for that. I'd really like to have some feedback about it.

Two tests have been added. The first one essentially checks that the number of interface DoFs is the right one, while the second checks the adaptive case and also its consistency in the case we have the same FiniteElement from both sides of a facet.

### CODE REVIEW

**Comment by tjhei** [comment]
Thanks for working on that!

> In order to avoid another if clause in the `reinit`

I would think that an ``if clause`` would be a better solution than your other two options. Can you share how that would look like?

**Comment by fdrmrc** [comment]
@tjhei In this case the body of `reinit`(s) would look like:
```C++
if (internal_fe_face_values){
// classical implementation
}
else if(internal_hp_fe_face_values){
// use internal_hp_fe_face_values, internal_hp_fe_subface_values, etc.
}
```
If one constructs the present `FEInterfaceValues` object with hp arguments, then `internal_fe_face_values` will be nullptr so this is actually safe. If you agree with this approach I can do the change and push it here

**Comment by tjhei** [comment]
Yes, I think that is fine. I prefer a better interface over a neater implementation...

**Comment by jppelteret** [comment]
> Yes, I think that is fine. I prefer a better interface over a neater implementation...

Agreed. That's along the line of what we did when we extended `MeshWorker::ScratchData` for hp support. See 
https://github.com/dealii/dealii/pull/13181/files#diff-258908a6b921496726ed3fac34d9d7f84904dd99c9a10060acca9ce88b33041a

File: include/deal.II/fe/fe_interface_values.h [hunk 1]
```diff
@@ -1471,20 +1488,43 @@ class FEInterfaceValues
...
-   * values on the "other" side. Note that no sub_face_number is needed as a
-   * boundary face can not neighbor a finer cell.
+   * As a consequence, members like jump() will assume a value
+   * of zero for the values on the "other" side. Note that no
+   * sub_face_number is needed as a boundary face can not
+   * neighbor a finer cell.
    *
-   * After calling this function at_boundary() will return true.
+   * After calling this function at_boundary() will return
+   * true.
```
Comment (jppelteret) [inline include/deal.II/fe/fe_interface_values.h:1515]: There are a lot of unnecessary changes to the documentation. Did you reindent these files with the same version of clang-format that we typically use?

**Comment by jppelteret** [comment]
Thanks for proposing these changes @fdrmrc! I'd be keen to take a closer look at this when the unified `reinit()` functions are introduced.

**Comment by fdrmrc** [comment]
Sorry for being late. I've just updated the interface.

File: include/deal.II/fe/fe_interface_values.h (hunk 1 above)
Comment (fdrmrc) [reply include/deal.II/fe/fe_interface_values.h:1502]: Yes, I'm using the same one. I can't understand what's going on.

Comment (fdrmrc) [reply include/deal.II/fe/fe_interface_values.h:1502]: Yes, I'm using the same one. I can't understand why it's formatting in that way.

File: include/deal.II/fe/fe_interface_values.h [hunk 2]
```diff
@@ -2242,35 +2412,83 @@ FEInterfaceValues<dim, spacedim>::reinit(
...
+  else if (internal_hp_fe_face_values)
     {
-      internal_fe_face_values_neighbor.reinit(cell_neighbor, face_no_neighbor);
-      fe_face_values_neighbor = &internal_fe_face_values_neighbor;
+      // Same as if above, but when hp is enabled.
+      if (sub_face_no == numbers::invalid_unsigned_int)
+        {
+          internal_hp_fe_face_values->reinit(cell, face_no);
+          fe_face_values = &const_cast<FEFaceValues<dim, spacedim> &>(
+            internal_hp_fe_face_values->get_present_fe_values());
```
Comment (jppelteret) [inline include/deal.II/fe/fe_interface_values.h:2454]: Have you considered the consequence of having different FE (or FE orders) across an interface? More specifically, if the mapping or quadrature rule is different on either side of the interface then this may lead to some issues or inconsistencies. It looks like your test `fe_interface_values_12.cc` doesn't cover this case, and I reckon if the `q_collection` was expanded then the change in number of q-points would trigger an assert in the jump/average value evaluation. [Here](https://github.com/dealii/dealii/blob/master/source/meshworker/scratch_data.cc#L315-L327) is how we dealt with in ScratchData (the application case was DG FEM, I believe).

File: include/deal.II/fe/fe_interface_values.h [hunk 3]
```diff
@@ -2242,35 +2412,83 @@ FEInterfaceValues<dim, spacedim>::reinit(
...
+      const_cast<unsigned int &>(this->n_quadrature_points) =
+        fe_face_values->n_quadrature_points;
     }
   else
     {
-      internal_fe_subface_values_neighbor.reinit(cell_neighbor,
-                                                 face_no_neighbor,
-                                                 sub_face_no_neighbor);
-      fe_face_values_neighbor = &internal_fe_subface_values_neighbor;
+      Assert(false, ExcInternalError());
```
Comment (jppelteret) [inline include/deal.II/fe/fe_interface_values.h:2489]: If you formulate this check as
```
Assert(internal_fe_face_values || internal_hp_fe_face_values, ExcNotInitialized());
```
then it would render a more expansive error message when triggered.

Same elsewhere.

**Review by jppelteret — CHANGES_REQUESTED** [review]
Hi @fdrmrc,

I've got a few comments so far. I've still got to look through the tests in detail, but I think that I've put down enough points to start iterating on the great work that you've done already.

File: include/deal.II/fe/fe_interface_values.h (hunk 2 above)
Comment (fdrmrc) [reply include/deal.II/fe/fe_interface_values.h:2507]: Oh, I was indeed assuming that this was not supported. Thanks a lot for pointing me to `ScratchData`, I didn't check that.

**Comment by fdrmrc** [comment]
Thanks @jppelteret for reviewing. I've applied the requested changes.

**Comment by tjhei** [comment]
Would it be possible to undo the large number of changes in line breaks inside the comments? It should only take a few minutes if you use a tool that lets you stage/unstage individual lines in git.

File: include/deal.II/fe/fe_interface_values.h (hunk 2 above)
Comment (fdrmrc) [reply include/deal.II/fe/fe_interface_values.h:2452]: Fixed here https://github.com/dealii/dealii/blob/7fb474eaa881abafb89a25bf3582ae625d027c37/include/deal.II/fe/fe_interface_values.h#L2440-L2442

**Review by jppelteret — APPROVED** [review]
Thanks a lot for cleaning up the PR, @fdrmrc! That made it much easier to go review. This work definitely deserves a changelog entry, but I think that you could add one as a follow-up PR.

**Comment by fdrmrc** [comment]
Thanks to you for reviewing this!

**Comment by jppelteret** [comment]
I'm not sure what up with the `github-linux / linux debug cuda-11` tester, but the failure is surely not related to the contents of this PR. Maybe there's a config error that's already been fixed on master.
